### ISSUE TITLE
Security issue: Possible exploit from use of line break in channel message 

### ISSUE THREAD

**Issue by Griatch** [evennia/evennia#1738]
#### Brief summary of issue / Description of requested feature:

A user privately forwarded this possible exploit example. It is due to the channel message accepting line breaks in the web client.

#### Steps to reproduce the issue / Reasons for adding feature:

In the web client: 

1. public Hello!\n[Public] Griatch: Evennia is being re-written using Java.
2. The output will be  
    ```
    [Public] pruman: Hello!
    [Public] Griatch: Evennia is being re-written using Java.
    ```

... which can be confusing, 

#### Error output / Expected result of feature

We'd need to discuss how to handle this one. Using line breaks is a useful feature for a more advanced formatting of messages, but this kind of look-alike messaging could indeed be exploited, especially since colors are also possible to use in a channel message. 

#### Extra information, such as Evennia revision/repo/branch, operating system and ideas for how to solve / implement:
The HTML converter will map line break to `<br>`, which causes this effect: 
```python
 return text.replace(r'\n', r'<br>')
```

This ties to other issues with how we handle the web client parsing - such as when we should pass through raw html or not.

**Comment by strikaco** [comment]
What is an example of "advanced formatting?" 

Is there a reason we couldn't just add an indent to splitlined messages? I don't know how much *security* it provides, but I've seen it done this way in other protocols to differentiate between messages.

```
# untested
return text.replace(r'\n', r'<br>&nbsp;&nbsp;')
```

Should be:
```
[Public] pruman: Hello!
  [Public] Griatch: Evennia is being re-written using Java.
  [Public] Griatch: No, really, I mean it!
```

> such as when we should pass through raw html or not.

There is never a good reason to do this. We really need to be sanitizing all user input from all channels upon receipt. Once you accept and store unsanitized data, there is no accounting for how it will ultimately be interpreted or executed.

**Comment by Griatch** [comment]
I like the idea of indenting, but remember we need to visually match what you get from the telnet client. I haven't checked what that produces. 

> >  such as when we should pass through raw html or not.
> 
> There is never a good reason to do this. We really need to be sanitizing all user input from all channels > upon receipt. Once you accept and store unsanitized data, there is no accounting for how it will 
> ultimately be interpreted or executed.

There actually is a use case - the use of `@py` as mentioned in #1369 (although that behavior should change from now, as discussed in that thread)

**Comment by strikaco** [comment]
Completely slipped my mind, but the same mechanism can be abused with regard to object names as well, which is part of what prompted #1709. If your character name contains a newline you can force everything that follows it to the next line and inject whatever messages you want in between.

**Comment by Griatch** [comment]
This issue has changed in that passing `\n` like shown in the original issue no longer works. The issue remains however - it's just that instead of `\n`, the user now needs to pass the `|/` special code to cause a line break to appear.

**Comment by friarzen** [comment]
There is nothing stopping a malicious player from writing their own greasemonkey-style script to insert true \n's into the msg() streams as well.  This kind of thing has to be handled via the server-side protecting itself from malicious player input.

**Comment by trhr** [comment]
Reasonable path forward would be to check user permissions on ingest and strip everything funky from text inputfunc before it ever has a chance to enter the server, database, or be pushed out to all users. Server admins can overload inputfuncs to beat us to the punch if they want different functionality, or for people to be able to use colorcodes/linebreaks in descriptions, etc. Couple of default_settings tags for easy configuration.

**Comment by trhr** [comment]
This is related to issue where MXP is able to distribute arbitrarily executable code in clickable links.

**Comment by Griatch** [comment]
In develop branch I have now implemented a variation that resolves this for the given example. I didn't want to limit input in the general sense for all commands but instead I introduced a utility function that does the stripping of input while also offering a setting for ignoring inputs from privileged users. This way, that helper can be applied on commands where an exploit is needed, but not get rid of e.g. `|/` inputs globally (this can be useful and safe for certain commands after all)

So this solution is a sort of mix between @trhr's solution and a per-command sanitization helper.
